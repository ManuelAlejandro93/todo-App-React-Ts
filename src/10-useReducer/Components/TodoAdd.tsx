import { useState } from 'react';

interface Props {
  onTODOAdd: Function;
}

export const TodoAdd = ({ onTODOAdd }: Props): JSX.Element => {
  const [input, setInput] = useState<string>('');

  const handleChange = (event: any): void => {
    const {
      target: { value }
    } = event;

    setInput(value);
  };

  const handleSubmit = (event: any): void => {
    event.preventDefault();
    if (input.trim().length < 6) return;
    onTODOAdd(input.trim());
    setInput('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        type='text'
        placeholder='Â¿QuÃ© hay que hacer?'
        className='form-control'
        value={input}
        onChange={handleChange}
      />
      <button
        type='submit'
        className='btn btn-outline-primary mt-2'
      >
        Agregar
      </button>
    </form>
  );
};
