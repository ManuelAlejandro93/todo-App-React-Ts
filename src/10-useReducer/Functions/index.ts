export { actionCreator } from './createActions';
export { initFn } from './initReducerFn';
export { updateLocalStorage } from './updateLocalStorage';
