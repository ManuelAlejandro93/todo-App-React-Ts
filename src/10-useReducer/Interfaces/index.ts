export type { Action } from './Action';
export type { UseToDoType, Todo } from './UseToDo';
