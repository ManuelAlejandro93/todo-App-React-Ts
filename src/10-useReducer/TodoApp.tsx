import { TodoList, TodoAdd } from './Components';
import { UseToDo } from './Hooks';

const TodoApp = (): JSX.Element => {
  const {
    todos,
    onAddTodo,
    onDeleteTodo,
    onToogleToDo,
    pendingToDos,
    totalToDos
  } = UseToDo();

  return (
    <>
      <h1>
        TODO App: {totalToDos} <small>Pendientes: {pendingToDos}</small>
      </h1>
      <hr />

      <div className='row'>
        <div className='col-7'>
          <TodoList
            todos={todos}
            deletefn={onDeleteTodo}
            onToogleToDo={onToogleToDo}
          />
        </div>

        <div className='col-5'>
          <h4>Agregar TODO</h4>
          <hr />

          <TodoAdd onTODOAdd={onAddTodo} />
        </div>
      </div>
    </>
  );
};

export default TodoApp;
