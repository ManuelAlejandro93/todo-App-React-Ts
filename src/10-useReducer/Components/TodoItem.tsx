import { Todo } from '../Interfaces';

interface Props {
  todo: Todo;
  deletefn: Function;
  onToogleToDo: Function;
}

export const TodoItem = ({
  todo,
  deletefn,
  onToogleToDo
}: Props): JSX.Element => {
  return (
    <li className='list-group-item d-flex justify-content-between custom-background mt-2'>
      <span
        className={`btn btn-primary ${todo.done ? 'line-through' : ''}`}
        onClick={() => onToogleToDo(todo.id)}
      >
        {todo.description}
      </span>

      <button
        className='btn btn-danger'
        onClick={() => deletefn(todo.id)}
      >
        Eliminar
      </button>
    </li>
  );
};
