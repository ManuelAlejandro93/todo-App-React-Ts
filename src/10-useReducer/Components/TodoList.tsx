import { Todo } from '../Interfaces';
import { TodoItem } from '../Components';

interface Props {
  todos: Todo[] | [];
  deletefn: Function;
  onToogleToDo: Function;
}
export const TodoList = ({
  todos,
  deletefn,
  onToogleToDo
}: Props): JSX.Element => {
  return (
    <ul className='list-group'>
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          deletefn={deletefn}
          onToogleToDo={onToogleToDo}
        />
      ))}
    </ul>
  );
};
