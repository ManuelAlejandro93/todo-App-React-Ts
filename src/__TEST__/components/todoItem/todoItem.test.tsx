import { fireEvent, render, screen } from '@testing-library/react';
import { TodoItem } from '@/10-useReducer/Components';

const todoMock = {
  id: 666,
  description: 'Todo para probar todoItem',
  done: false
};

describe('Pruebas en <TodoItem>', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Crear un snapshot', () => {
    const delefnMock = jest.fn(() => {});
    const onToogleToDoMock = jest.fn(() => {});

    const { container } = render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    expect(container).toMatchSnapshot();
  });
  test('Al dar click en el boton eliminar, 4 veces. Se debe llamar 4 veces a la función deletefn', () => {
    const delefnMock = jest.fn(() => {});
    const onToogleToDoMock = jest.fn(() => {});

    render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const deleteButton = screen.getByText('Eliminar');

    fireEvent.click(deleteButton);
    fireEvent.click(deleteButton);
    fireEvent.click(deleteButton);
    fireEvent.click(deleteButton);

    expect(delefnMock).toHaveBeenCalledTimes(4);
  });

  test('Al dar click en el boton eliminar, se debe llamar la funcion delefnMock con el valor del todo id.', () => {
    const delefnMock = jest.fn();
    const onToogleToDoMock = jest.fn();

    render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const deleteButton = screen.getByText('Eliminar');

    fireEvent.click(deleteButton);

    expect(delefnMock).toHaveBeenCalledWith(666);
  });

  test('Al dar click en el boton toogleTodo, 4 veces. Se debe llamar 4 veces a la función ontoggleTodo con el valor 666', () => {
    const delefnMock = jest.fn();
    const onToogleToDoMock = jest.fn();

    const { container } = render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const spanToggleElement = container.querySelector('span');

    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);

    expect(onToogleToDoMock).toHaveBeenNthCalledWith(4, 666);
  });

  test('Al dar click en el boton toogleTodo, 4 veces. Se debe llamar 4 veces a la función ontoggleTodo', () => {
    const delefnMock = jest.fn(() => {});
    const onToogleToDoMock = jest.fn(() => {});

    const { container } = render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const spanToggleElement = container.querySelector('span');

    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);
    fireEvent.click(spanToggleElement as HTMLElement);

    expect(onToogleToDoMock).toHaveBeenCalledTimes(4);
  });

  test('Al renderizar debe existir en pantalla el elemento "Todo para probar todoItem"', () => {
    const delefnMock = jest.fn();
    const onToogleToDoMock = jest.fn();

    render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const todoButton = screen.getByText(/Todo para probar todoItem/i);
    expect(todoButton).toBeTruthy();
  });

  test('Al dar click en el span con texto: "Todo para probar todoItem", NO debe cambiar la clase de css para incluir el texto: "line-through". Porque todo es un Mock.', () => {
    const delefnMock = jest.fn();
    const onToogleToDoMock = jest.fn();

    render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const todoButton = screen.getByText(/Todo para probar todoItem/i);
    expect(todoButton).toBeTruthy();

    fireEvent.click(todoButton);

    expect(todoButton.classList.contains('line-through')).toBeFalsy();
  });

  test('Al enviar un to-do completado debe renderizar y tener la clase "line-through" que tacha la tarea.', () => {
    const delefnMock = jest.fn();
    const onToogleToDoMock = jest.fn();

    todoMock.done = true;

    render(
      <TodoItem
        deletefn={delefnMock}
        onToogleToDo={onToogleToDoMock}
        key={'1'}
        todo={todoMock}
      />
    );

    const todoButton = screen.getByText(/Todo para probar todoItem/i);
    expect(todoButton).toBeTruthy();

    fireEvent.click(todoButton);

    expect(todoButton.classList.contains('line-through')).toBeTruthy();
  });
});
