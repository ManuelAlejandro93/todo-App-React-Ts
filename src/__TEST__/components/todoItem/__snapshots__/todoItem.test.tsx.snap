// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoItem> Crear un snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      class="btn btn-primary "
    >
      Todo para probar todoItem
    </span>
    <button
      class="btn btn-danger"
    >
      Eliminar
    </button>
  </li>
</div>
`;
